src/Features/Core/Portable/ExternalAccess/UnitTesting/API/UnitTestingStackTraceService.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.StackTraceExplorer;
+using System.Collections.Immutable;
+using System;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.UnitTesting.API
+{
+    internal sealed class UnitTestingStackTraceService
+    {
+        private readonly IStackTraceExplorerService _stackTraceService;
+
+        public UnitTestingStackTraceService(HostWorkspaceServices services)
+            => _stackTraceService = services.GetRequiredService<IStackTraceExplorerService>();
+
+        public async Task<Frame?> TryParseAsync(string line, CancellationToken cancellationToken)
+        {
+            var result = await StackTraceAnalyzer.AnalyzeAsync(line, cancellationToken).ConfigureAwait(false);
+
+            if (result.ParsedFrames.Length == 0)
+            {
+                return null;
+            }
+
+            if (result.ParsedFrames.Length > 1)
+            {
+                throw new InvalidOperationException();
```

Comment:
Why not expose an API to parse an entire stack trace and return `IEnumerable<Frame>` (in addition to the current per-frame API)?

I haven't looked into exactly how we will consume this on UT side. But for completeness it would be nice to have an API that supports parsing an entire trace (especially since the internal API supports this).

Reply:
Whatever works best for the UT team. We can do either. This was my guess at "best" way to do it so you handle line split and we parse a single line. If you want the whole trace to be handled by us then IEnumerable makes sense

Reply:
Its not a huge issue but I think handling the whole trace in here may be better since that doesn't preclude passing in individual lines.

In other words, with an API that returns IEnumerable, I can still pass in individual lines and call .Single on the retuned enumerable on our side. However, the opposite is not possible (i.e. I can never pass in an entire stack trace if the API only supports passing one at a time).